### PR TITLE
NFC: Provide best practices for errors

### PR DESCRIPTION
**Pull request by TedDriggs**
This improves the documentation about how best to use `darling::Error`

### CODE REVIEW

**Comment by TedDriggs** [comment]
@chipsenkbeil can you take a look at this?

**Comment by TedDriggs** [comment]
@chipsenkbeil correct. `darling` errors bind to the tightest span they can, so choosing one of them is likely to point to some random field or meta-item, rather than to any usefully-broad span. Using `Span::call_site()` is the appropriate move for that, though as seen in #110 that will cause the error to point at the macro invocation.

Having looked at `entity-rs`, I don't see it doing anything that precludes the use of `darling::Error` for its custom validations after initial parsing. If there is such a place, can you link me to it? I'd be willing to add a conversion from `syn::Error` into `darling::Error` if you're getting `syn` errors back from the crate as opposed to making them in `entity-rs`.

File: core/src/error/mod.rs [hunk 1]
```diff
@@ -26,6 +26,35 @@ pub type Result<T> = ::std::result::Result<T, Error>;
...
+///    create a local `Vec` to collect errors as they are encountered and then use
+///    `darling::Error::multiple` to create an error containing all those issues if the list
+///    is non-empty after validation. This can create very complex custom validation functions;
+///    in those cases, split independent "validation chains" out into their own functions to
+///    keep the main validator manageable.
+/// 3. Use `darling::Error::custom` to create additional errors as-needed, then call `with_span`
+///    to ensure those errors appear in the right place. Use `darling::util::SpannedValue` to keep
+///    span information around on parsed fields so that custom diagnostics can point to the correct
+///    parts of the input AST.
+/// 4. Enable the `diagnostics` feature of `darling`.
```
Comment (TedDriggs) [inline core/src/error/mod.rs:57]: This still requires nightly, so this isn't a good idea.

```suggestion
```

**Comment by chipsenkbeil** [comment]
> @chipsenkbeil correct. `darling` errors bind to the tightest span they can, so choosing one of them is likely to point to some random field or meta-item, rather than to any usefully-broad span. Using `Span::call_site()` is the appropriate move for that, though as seen in #110 that will cause the error to point at the macro invocation.
> 
> Having looked at `entity-rs`, I don't see it doing anything that precludes the use of `darling::Error` for its custom validations after initial parsing. I did see some places where you use the `?` operator a function that returns a `Result<_, syn::Error>`, so I've added a conversion from `syn::Error` to `darling::Error` to make that seamless in #116

I'm planning to do a little refactoring to clean up the code and split out the derive into multiple derives, so I can work on switching over to `darling::Error` at the same time. Thanks for all of the help and patience!

**Pull request merged**